File: webapp/js/services/budgetService.js
import { hostRequest } from '../core/hostBridge.js';

function toNullableNumber(value) {
  if (value === '' || value == null) {
    return null;
  }

  const parsed = Number(value);
  return Number.isNaN(parsed) ? null : parsed;
}

function normalizeRows(rows) {
  return (Array.isArray(rows) ? rows : []).map((row) => ({
    budgetRowId: row.budgetRowId ?? null,
    deptName: String(row.deptName || '').trim(),
    plannedQty: toNullableNumber(row.plannedQty),
    usedQty: toNullableNumber(row.usedQty),
    reasonText: String(row.reasonText || '')
  }));
}

export const budgetService = {
  async loadBudget(sessionId, userName = '') {
    return hostRequest('budget.load', { sessionId, userName });
  },

  async saveBudget(sessionId, rows, meta = null, userName = '') {
    return hostRequest('budget.save', {
      sessionId,
      userName,
      rows: normalizeRows(rows),
      meta
    });
  },

  async recalculate(sessionId, rows, meta = null) {
    return hostRequest('budget.recalculate', {
      sessionId,
      rows: normalizeRows(rows),
      meta
    });
  },

  async loadDashboardBudgetSummary(sessionId) {
    return hostRequest('dashboard.budgetSummary', { sessionId });
  }
};

File: desktop-host/src/BudgetService.cs
using System;
using System.Collections.Generic;

namespace MoatHouseHandover.Host;

public sealed class BudgetService
{
    private readonly BudgetRepository _repository;
    private readonly AuditLogService _auditLogService;

    public BudgetService(BudgetRepository repository, AuditLogService auditLogService)
    {
        _repository = repository;
        _auditLogService = auditLogService;
    }

    public BudgetPayload LoadBudget(BudgetLoadRequest request)
    {
        if (request.SessionId <= 0)
        {
            throw new InvalidOperationException("SessionId is required.");
        }

        return _repository.LoadBudget(request.SessionId, NormalizeUser(request.UserName));
    }

    public BudgetPayload SaveBudget(BudgetSaveRequest request)
    {
        if (request.SessionId <= 0)
        {
            throw new InvalidOperationException("SessionId is required.");
        }

        var rows = request.Rows ?? Array.Empty<BudgetRowUpsertRequest>();
        ValidateRows(rows);
        var userName = NormalizeUser(request.UserName);
        var payload = _repository.SaveBudget(request.SessionId, rows, userName);
        _auditLogService.BestEffortLog(
            actionType: "budget.save",
            entityType: "Budget",
            entityKey: AuditLogService.BuildSessionKey(request.SessionId),
            userName: userName,
            details: new { sessionId = request.SessionId, rowCount = rows.Count });

        return payload;
    }

    public BudgetPayload Recalculate(BudgetRecalculateRequest request)
    {
        if (request.SessionId <= 0)
        {
            throw new InvalidOperationException("SessionId is required.");
        }

        var rows = request.Rows ?? Array.Empty<BudgetRowUpsertRequest>();
        ValidateRows(rows);
        return _repository.Recalculate(request.SessionId, rows);
    }

    public BudgetSummaryPayload LoadBudgetSummary(long sessionId)
    {
        if (sessionId <= 0)
        {
            throw new InvalidOperationException("SessionId is required.");
        }

        return _repository.LoadBudgetSummary(sessionId);
    }

    private static void ValidateRows(IReadOnlyList<BudgetRowUpsertRequest> rows)
    {
        for (var i = 0; i < rows.Count; i += 1)
        {
            var row = rows[i];
            if (string.IsNullOrWhiteSpace(row.DeptName))
            {
                throw new InvalidOperationException($"Budget row {i + 1}: deptName is required.");
            }

            if (row.PlannedQty.HasValue && row.PlannedQty.Value < 0)
            {
                throw new InvalidOperationException($"Budget row '{row.DeptName}': planned value cannot be negative.");
            }

            if (row.UsedQty.HasValue && row.UsedQty.Value < 0)
            {
                throw new InvalidOperationException($"Budget row '{row.DeptName}': used value cannot be negative.");
            }
        }
    }

    private static string NormalizeUser(string userName)
    {
        if (!string.IsNullOrWhiteSpace(userName))
        {
            return userName.Trim();
        }

        return Environment.UserName;
    }
}
